### PR TITLE
feat: introduce new package @commercetools-backend/express with utilities for working with Custom Applications

### PR DESCRIPTION
**Pull request by emmenko**
Ref #1444 

cc @adnasa for awareness

### CODE REVIEW

File: packages/server-express/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+> You can read more about the "Proxy to External API" concepts [here](https://docs.commercetools.com/custom-applications/main-concepts/proxy-to-external-api).
+
+```js
+const {
+  createSessionMiddleware,
+  CLOUD_IDENTIFIERS,
+} = require('@commercetools-frontend/server-express');
+
+app.use(createSessionMiddleware({ mcApiUrl: CLOUD_IDENTIFIERS['gcp-eu'] }));
```
Comment (tdeekens) [inline packages/server-express/README.md:29]: Would `CLOUD_IDENTIFIERS.GCP_EU` also work? Just as it's easier to type.

File: packages/server-express/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+- `issuer` (_string_): either a cloud identifier or a URL. The cloud identifier maps to the MC API URL of the related [cloud region](https://docs.commercetools.com/custom-applications/main-concepts/api-gateway#cloud-regions).
+
+  - `gcp-au`: `https://mc-api.australia-southeast1.gcp.commercetools.com`
+  - `gcp-eu`: `https://mc-api.europe-west1.gcp.commercetools.com`
+  - `gcp-us`: `https://mc-api.us-central1.gcp.commercetools.com`
+  - `aws-fra`: `https://mc-api.eu-central-1.aws.commercetools.com`
+  - `aws-ohio`: `https://mc-api.us-east-2.aws.commercetools.com`
+
+- `inferIssuerFromCustomHeader` (_boolean_): determines whether the issuer should be inferred from the custom request HTTP header `x-mc-api-cloud-identifier` which is sent by the MC API when forwarding the request. This might be useful in case the server is used in multiple regions.
```
Comment (tdeekens) [inline packages/server-express/README.md:45]: Maybe we can just say `inferIssuer`. What the middleware used internally: right now the header then it's up to the middleware?

File: packages/server-express/src/middlewares/fixtures/jwt-token.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { JWT, JWK, JWKS } from '@panva/jose';
+
+const keyRS256 = JWK.generateSync('RSA', 2048, { use: 'sig', alg: 'RS256' });
+
+export const jwksStore = new JWKS.KeyStore([keyRS256]);
+
+export const createToken = (options: { issuer: string; audience: string }) =>
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/fixtures/jwt-token.ts:7]: Nit: exporting in the end `exports { jwksStore, createToken }` helps (at least me) when scanning the code :)

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import jwksRsa from 'jwks-rsa';
+
+const decodedTokenKey = 'decoded_token';
+
+const mapCloudIdentifierToUrl = (
+  issuer: TSessionMiddlewareOptions['issuer']
+) => {
+  switch (issuer) {
+    case CLOUD_IDENTIFIERS['gcp-au']:
+      return 'https://mc-api.australia-southeast1.gcp.commercetools.com';
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:15]: Should we keep these constants in a module in this package of maybe even outside?

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import { CLOUD_IDENTIFIERS } from '../constants';
+import expressJwtMiddleware from 'express-jwt';
+import jwksRsa from 'jwks-rsa';
+
+const decodedTokenKey = 'decoded_token';
+
+const mapCloudIdentifierToUrl = (
+  issuer: TSessionMiddlewareOptions['issuer']
+) => {
+  switch (issuer) {
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:13]: Sorry, got confused here. We map the cloud identifier to an url but we use the issuer to do so. Should we name the issuer just cloudIdentifier?

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  request.session = {
+    userId: decodedToken.sub,
+    projectKey: decodedToken[publicClaimForProjectKey],
+  };
+
+  delete request.decoded_token;
+};
+
+function createSessionMiddleware(options: TSessionMiddlewareOptions) {
+  const configuredIssuer =
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:53]:
```suggestion
  const configuredDefaultIssuer =
```

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    mapCloudIdentifierToUrl(options.issuer) ?? options.issuer;
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:54]: Can `mapCloudIdentifierToUrl` not return the `options.issuer` if it fails to match? We nullish into it.

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  // Return the middleware
+  return (request: Request, response: Response, next: NextFunction) => {
+    let issuer = configuredIssuer;
+    if (options.inferIssuerFromCustomHeader) {
+      // Attempt to infer the
+      const cloudIdentifierHeader = request.header('x-mc-api-cloud-identifier');
+      if (cloudIdentifierHeader) {
+        issuer = mapCloudIdentifierToUrl(cloudIdentifierHeader) ?? issuer;
+      }
+    }
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:65]:
```suggestion
    const cloudIdentifierHeader = request.header('x-mc-api-cloud-identifier');
    const overwriteDefaultIssuer = options.inferIssuerFromCustomHeader && cloudIdentifierHeader
    const issuer = overwriteDefaultIssuer ? mapCloudIdentifierToUrl(cloudIdentifierHeader) ?? issuer : configuredDefaultIssuer;
```

Would help be get around this if nesting maybe somehow.

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  }
+};
+
+type TDecodedJWT = {
+  sub: string;
+  iss: string;
+  [property: string]: string;
+};
+
+const writeSessionContext = (
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:37]:
```suggestion
const exposeSessionOnRequest = (
```

File: packages/server-express/src/middlewares/session-middleware.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      // Validate the audience and the issuer.
+      audience,
+      issuer,
+      algorithms: ['RS256'],
+    })(request, response, (error) => {
+      if (error) {
+        return next(error);
+      }
+      // @ts-ignore
+      writeSessionContext(request);
```
Comment (tdeekens) [inline packages/server-express/src/middlewares/session-middleware.ts:86]:
```suggestion
      exposeSessionOnRequest(request);
```

Comment (tdeekens) [reply packages/server-express/src/middlewares/session-middleware.ts:86]: 🤷

File: packages/server-express/src/middlewares/session-middleware.ts (hunk 4 above)
Comment (emmenko) [reply packages/server-express/src/middlewares/session-middleware.ts:15]: Maybe later, not sure yet how things will evolve here.

File: packages/server-express/src/middlewares/session-middleware.ts (hunk 5 above)
Comment (emmenko) [reply packages/server-express/src/middlewares/session-middleware.ts:13]: I went back and forth with this and eventually settled on this: the `issuer` is a required field, and you can pass one of the identifiers. However you can also "bail out" and pass the URL yourself.
Initially I had it split into two fields, but both are then optional and it just gets complicated for validation and stuff.

I'm all hears if you have other suggestions.

File: packages/server-express/src/middlewares/session-middleware.ts (hunk 8 above)
Comment (emmenko) [reply packages/server-express/src/middlewares/session-middleware.ts:65]: Hmm I find the nesting more explicit and easier to read. Also it helps TS to infer things better.

File: packages-backend/server-express/README.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
+# @commercetools-backend/server-express
```
Comment (tdeekens) [inline packages-backend/server-express/README.md:1]: Sorry, do we do `express` or `server-express`? Intuitively I would just go with `express` 🤷 😕

File: packages-backend/server-express/src/middlewares/session-middleware.ts [hunk 12]
```diff
@@ -50,27 +59,33 @@ const writeSessionContext = (
...
-    if (options.inferIssuerFromCustomHeader) {
-      // Attempt to infer the
-      const cloudIdentifierHeader = request.header('x-mc-api-cloud-identifier');
-      if (cloudIdentifierHeader) {
-        issuer = mapCloudIdentifierToUrl(cloudIdentifierHeader) ?? issuer;
-      }
-    }
+    const cloudIdentifierHeader = request.header('x-mc-api-cloud-identifier');
+    const issuer =
+      options.inferIssuer && cloudIdentifierHeader
```
Comment (tdeekens) [inline packages-backend/server-express/src/middlewares/session-middleware.ts:72]: Danke 🖤!

File: packages-backend/server-express/src/middlewares/session-middleware.ts [hunk 13]
```diff
@@ -1,32 +1,41 @@
...
-      return 'https://mc-api.us-east-2.aws.commercetools.com';
+    case CLOUD_IDENTIFIERS.AWS_FRA:
+      return MC_API_URLS.AWS_FRA;
+    case CLOUD_IDENTIFIERS.AWS_OHIO:
+      return MC_API_URLS.AWS_OHIO;
     default:
       return undefined;
   }
 };
+const throwIfIssuerIsNotAValidUrl = (issuer: string) => {
```
Comment (tdeekens) [inline packages-backend/server-express/src/middlewares/session-middleware.ts:30]: 🏨

File: packages-backend/server-express/README.md (hunk 11 above)
Comment (emmenko) [reply packages-backend/server-express/README.md:1]: Yeah I guess with the "backend" prefix we can just use `express`

**Comment by emmenko** [comment]
So for further work on the package, I suppose we can merge this and open other PRs?

**Comment by emmenko** [comment]
Just a mistake 😅

**Pull request merged**